sql: resolve public.geometry and public.geography as built-in types

otan opened the pull request:
To maintain compatability with PostGIS (especially backups), we need to
be able to resolve public.geometry and public.geography as the built-in
GEOMETRY and GEOGRAPHY types. Add that ability here.

Also prevent creating a type public.geometry and public.geography to
reduce the possible conflicts.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52409)
<!-- Reviewable:end -->

rohany commented on pkg/sql/create_type.go:
```diff
@@ -78,6 +78,14 @@ func resolveNewTypeName(
 ) (typeKey sqlbase.DescriptorKey, schemaID sqlbase.ID, err error) {
+	// Account for geometry and geography being a PostgreSQL extension
+	// whose types live in the public schema.
+	if name.Schema() == string(tree.PublicSchemaName) {
```
nit can just do `tree.PublicSchema`

rohany commented on pkg/sql/sem/tree/type_name.go:
```diff
@@ -131,6 +131,17 @@ func ResolveType(
 	case *UnresolvedObjectName:
+		// Account for geometry and geography being a PostgreSQL extension
+		// whose types live in the public schema.
+		if t.Schema() == string(PublicSchemaName) {
```
I wouldn't put this here -- I think this should go in the `LogicalSchemaAccessor`

rohany commented on pkg/sql/sem/tree/type_name.go:
```diff
@@ -131,6 +131,17 @@ func ResolveType(
+		// whose types live in the public schema.
+		if t.Schema() == string(PublicSchemaName) {
+			switch t.Object() {
+			case "geometry":
```
are there constants for these?

otan replied:
nothing existing

otan replied:
did something more interesting

jordanlewis commented:
drive-by nit: s/native/built-in/ or another more precise word (https://developers.google.com/style/word-list#native)

rohany commented on pkg/sql/catalog/accessors/logical_schema_accessors.go:
```diff
@@ -111,6 +113,16 @@ func (l *LogicalSchemaAccessor) GetObjectDesc(
 		return desc.Desc(), nil
 	}
+
+	if schema == tree.PublicSchema {
```
And check that we are looking for a type in the flags

rohany commented on pkg/sql/create_type.go:
```diff
@@ -78,6 +78,11 @@ func resolveNewTypeName(
 	params runParams, name *tree.TypeName, db *sqlbase.ImmutableDatabaseDescriptor,
 ) (typeKey sqlbase.DescriptorKey, schemaID sqlbase.ID, err error) {
+	if name.Schema() == tree.PublicSchema {
+		if _, ok := types.PublicSchemaAliases[name.Object()]; ok {
```
fancy

need to also do the same in getCreateTableParams

otan replied:
we can already add tables with type names atm (though i'm sure that's not intentional)

```
root@127.0.0.1:59943/defaultdb> create table int4 (a int);
CREATE TABLE

Time: 7.211ms
```

i'd rather save that change for something more wide-catching.

otan replied:
ill add a TODO i guess :P

otan replied:
ah nvm i'm dumb!

rohany commented on pkg/sql/logictest/testdata/logic_test/geospatial:
```diff
@@ -37,6 +37,29 @@ CREATE TABLE geom_table_negative_values(
+CREATE TYPE geometry AS enum('no')
+
+statement error type .*geography.* already exists
+CREATE TYPE geography AS enum('no')
```
and add a create table test

rohany commented on pkg/sql/catalog/accessors/logical_schema_accessors.go:
```diff
@@ -112,6 +114,16 @@ func (l *LogicalSchemaAccessor) GetObjectDesc(
 		return desc.Desc(), nil
 	}
+
+	if schema == tree.PublicSchema && flags.DesiredObjectKind == tree.TypeObject {
```
I'd just add some comments here and in the create*Params about what is going on

rohany commented on pkg/sql/types/alias.go:
```diff
@@ -0,0 +1,19 @@
+// Copyright 2015 The Cockroach Authors.
```
2020

otan replied:
nah mate i'm pretty sure it's 2015, no corona

otan committed:
sql: resolve public.geometry and public.geography as built-in types

To maintain compatability with PostGIS (especially backups), we need to
be able to resolve public.geometry and public.geography as the built-in
GEOMETRY and GEOGRAPHY types. Add that ability here.

Also prevent creating a type public.geometry and public.geography to
reduce the possible conflicts.

Release note: None

otan commented:
bors r=rohany
